src/runtime/virtcontainers/nydus.go

```diff
@@ -0,0 +1,350 @@
+	"os/exec"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/kata-containers/kata-containers/src/runtime/pkg/katautils/katatrace"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	infoEndpoint  = "/api/v1/daemon"
+	mountEndpoint = "/api/v1/mount"
+
+	defaultHttpClientTimeout = 30 * time.Second
+	contentType              = "application/json"
+
+	nydusRafsType          = "rafs"
+	nydusPassthroughfsType = "passthrough_fs"
+
+	configPrefix   = "config="
+	snapshotPrefix = "snapshotdir="
+
+	sharedPathInGuest = "/containers"
+)
+
+var nydusdTracingTags = map[string]string{
+	"source":    "runtime",
+	"package":   "virtcontainers",
+	"subsystem": "nydusd",
+}
+
+type nydusd struct {
+	pid         int
+	path        string
+	sockPath    string
+	apiSockPath string
+	sourcePath  string
+	logFile     string
+	extraArgs   []string
+}
+
+func (nd *nydusd) Start(ctx context.Context, onQuit onQuitFunc) (int, error) {
+	span, _ := katatrace.Trace(ctx, nd.Logger(), "Start", nydusdTracingTags)
+	defer span.End()
+	pid := 0
+
+	args := []string{
+		"--log-level", "debug",
+		"--log-file", nd.logFile,
+		"--apisock", nd.apiSockPath,
+		"--sock", nd.sockPath,
+	}
+	if len(nd.extraArgs) > 0 {
+		args = append(args, nd.extraArgs...)
+	}
+	cmd := exec.Command(nd.path, args...)
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return pid, err
+	}
+	nd.Logger().WithField("path", nd.path).Info()
+	nd.Logger().WithField("args", strings.Join(args, " ")).Info()
+	if err := cmd.Start(); err != nil {
+		return pid, err
+	}
+	// Monitor virtiofsd's stderr and stop sandbox if virtiofsd quits
+	go func() {
+		scanner := bufio.NewScanner(stderr)
+		for scanner.Scan() {
+			nd.Logger().WithField("source", "nydusd").Info(scanner.Text())
+		}
+		nd.Logger().Info("nydusd quits")
+		// Wait to release resources of virtiofsd process
+		cmd.Process.Wait()
+		if onQuit != nil {
+			onQuit()
+		}
+	}()
+	if err := nd.setupPassthroughFS(); err != nil {
+		return pid, err
+	}
+	nd.pid = cmd.Process.Pid
+	return nd.pid, nil
+}
+
+func (nd *nydusd) setupPassthroughFS() error {
+	if err := os.MkdirAll(nd.sourcePath, DirMode); err != nil {
+		return err
+	}
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("from", nd.sourcePath).
+		WithField("dest", sharedPathInGuest).Info("prepare mount passthroughfs")
+
+	mr := NewMountRequest(nydusPassthroughfsType, nd.sourcePath, "")
+	return nc.Mount(sharedPathInGuest, mr)
+}
+
+func (nd *nydusd) MountRAFS(opt MountOption) error {
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("from", opt.bootstrap).
+		WithField("dest", opt.mountpoint).Info("prepare mount rafs")
+
+	mr := NewMountRequest(nydusRafsType, opt.bootstrap, opt.daemonConfig)
+	return nc.Mount(opt.mountpoint, mr)
+}
+
+func (nd *nydusd) Stop(ctx context.Context) error {
+	if err := nd.kill(ctx); err != nil {
+		return nil
```

Comment:
Here should print the err to log. `virtiofsd.go` has the same problem, can you update it by the way in this PR ?